### DISCUSSION TITLE
use of tolerances

### DISCUSSION THREAD

**Discussion by lkapelevich** [ERGO-Code/HiGHS#1457] [General]
I noticed that when passing an initial solution for a MIP, the `primal_feasibility_tolerance` is utilized in the feasibility check. However, only `mip_feasibility_tolerance` is used for checking solutions when the solver is running. I have two questions/comments:
1) This means it's possible to grab a solution out of HiGHS that is deemed feasible but when it is passed as an initial solution to the same problem, it's deemed infeasible
2) Do you know of other situations where either `primal_feasibility_tolerance` or `dual_feasibility_tolerance` values are used if solving a problem with integrality? Or, is it safe to assume that only `mip_feasibility_tolerance` matters?

**Comment by jajhall** [comment]
Well spotted! 

Since the default value of `primal_feasibility_tolerance` is 1e-7, and the default value of `mip_feasibility_tolerance` is 1e-6, any solution that's deemed to be a feasible initial solution before calling the MIP solver will be feasible when checked within the MIP solver - _unless_ a user modifies the tolerances so that `mip_feasibility_tolerance` < `primal_feasibility_tolerance`. I should resolve this, but it's not a priority.

1. So, yes, this is possible in theory. However, what user is going to modify the tolerances in this way?
2. No, `primal_feasibility_tolerance` and `dual_feasibility_tolerance` are not used when solving a MIP, so only `mip_feasibility_tolerance` matters

**Reply by lkapelevich** [reply]
So I think we can run into this with the defaults `primal_feasibility_tolerance=1e-7` and `mip_feasibility_tolerance =1e-6`. For example, if I set up the problem:
min 0 : 
1 x <= -1e-6
x \in [0, inf]
x integer
I get the solution x = 0. If I set that as a starting value, it's deemed infeasible.

**Reply by jajhall** [reply]
I didn't think that you were referring to this situation - but now I see you're referring to a solution _out of_ HiGHS 

I'll fix this in due course

**Reply by lkapelevich** [reply]
Sounds good. And thanks for clarifying `mip_feasibility_tolerance`.

**Reply by lkapelevich** [reply]
> No, `primal_feasibility_tolerance` and `dual_feasibility_tolerance` are not used when solving a MIP, so only `mip_feasibility_tolerance` matters

Maybe one thing just to make sure: this includes any tolerances used in pre/post processing, right?

**Reply by jajhall** [reply]
I can't say. Although there are no meaningful references to primal_feasibility_tolerance in presolve and postsolve, there are sure to be some magic numbers there.

**Reply by fwesselm** [reply]
It seems that `primal_feasibility_tolerance` is used in `HighsPostsolveStack::DuplicateColumn::undo` and related code.

For example, shouldn't `mip_feasibility_tolerance` (or its value) be used [here](https://github.com/ERGO-Code/HiGHS/blob/624bc58d3f2fdbc7f0c1e279da4d46779d04a13e/src/presolve/HighsPostsolveStack.cpp#L577) if a MIP solution is postsolved?

**Reply by jajhall** [reply]
When I first looked at postsolve, I thought that the use of `primal_feasibility_tolerance` was just for some debugging code that I introduced, but you're right. Sometime I'll look through the use of tolerances in presolve/postsolve as it could explain some very occasional strange behaviour, but I can't prioritise it
